torch/csrc/jit/frontend/ir_emitter.cpp

```diff
@@ -3689,24 +3754,79 @@ struct to_ir {
         }
         AT_ASSERT(key_type != nullptr && value_type != nullptr);
 
-        auto checkTypeOfValues = [](const TypePtr& type,
-                                    const char* what,
-                                    const std::vector<Value*>& values,
-                                    TreeList trees) {
-          for (size_t i = 0, N = values.size(); i < N; ++i) {
-            std::stringstream ss;
-            if (!values[i]->type()->isSubtypeOfExt(type, &ss)) {
-              throw ErrorReport(trees[i])
-                  << "Dict " << what
-                  << " must contain only a single type, expected: "
-                  << type->repr_str() << " but found "
-                  << values[i]->type()->repr_str() << " instead.\n"
-                  << ss.str();
+        for (size_t i = 0, N = keys.size(); i < N; ++i) {
+          std::stringstream ss;
+          if (!keys[i]->type()->isSubtypeOfExt(key_type, &ss)) {
+            throw ErrorReport(key_trees[i])
+                << "Dict keys must contain "
+                << "only a single type, expected: " << key_type->repr_str()
+                << " but found " << keys[i]->type()->repr_str() << " instead.\n"
+                << ss.str();
+          }
+        }
+
+        if (!values.empty()) {
+          // Get the smallest common supertype of all the values. We
+          // need to use the O(N^2) algorithm instead of using
+          // `std::accumulate` to account for the case that two types
+          // cannot themselves be unified, but there exists some mutual
+          // parent type later in `values`. (E.g. T2 <: T1, T3 <: T1,
+          // `values` = {T2, T3, T1}. The real supertype should be `T1`,
+          // not `Any`)
+
+          // List of possible parent types
+          std::vector<TypePtr> seen;
+
+          auto get_greater_type =
+              [](const TypePtr t1, const TypePtr t2) -> c10::optional<TypePtr> {
+            if (t1->isSubtypeOf(t2)) {
+              return t2;
+            } else if (t2->isSubtypeOf(t1)) {
+              return t1;
+            } else {
+              return c10::nullopt;
+            }
+          };
```

Comment:
This looks very similar to `unifyTypes(t1, t2, /*default_to_any=*false/)`, could you clarify the differences?